Exception in DepInvalidator

Steps to reproduce:
1. Checkout https://github.com/koperagen/npe-repro
2. Execute task `use-preprocessor:kspKotlin`
3. Comment out "@MyAnnotation" in use-preprocessor org.example.Test, so it's like this
```
//@MyAnnotation
class Test
```
4. Execute task `use-preprocessor:kspKotlin`
5. Uncomment the annotation
```
@MyAnnotation
class Test
```
6. Execute task `use-preprocessor:kspKotlin`. At this point exception is thrown:

```
java.lang.IllegalStateException: failed to analyze: java.lang.NullPointerException
	at org.jetbrains.kotlin.analyzer.AnalysisResult.throwIfError(AnalysisResult.kt:56)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:105)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli$default(KotlinToJVMBytecodeCompiler.kt:58)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:170)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:52)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:92)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:44)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:98)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1618)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:359)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:562)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:796)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:677)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:676)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at com.google.devtools.ksp.DepInvalidator.invalidate(Incremental.kt:798)
	at com.google.devtools.ksp.DepInvalidator.invalidate(Incremental.kt:790)
	at com.google.devtools.ksp.IncrementalContext.calcDirtySetByDeps(Incremental.kt:290)
	at com.google.devtools.ksp.IncrementalContext.calcDirtyFiles(Incremental.kt:436)
	at com.google.devtools.ksp.AbstractKotlinSymbolProcessingExtension$doAnalysis$1.invoke(KotlinSymbolProcessingExtension.kt:144)
	at com.google.devtools.ksp.AbstractKotlinSymbolProcessingExtension$doAnalysis$1.invoke(KotlinSymbolProcessingExtension.kt:138)
	at com.google.devtools.ksp.AbstractKotlinSymbolProcessingExtension.handleException(KotlinSymbolProcessingExtension.kt:287)
	at com.google.devtools.ksp.AbstractKotlinSymbolProcessingExtension.doAnalysis(KotlinSymbolProcessingExtension.kt:138)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration(TopDownAnalyzerFacadeForJVM.kt:120)
	at org.jetbrains.kotlin.cli.jvm.compiler.TopDownAnalyzerFacadeForJVM.analyzeFilesWithJavaIntegration$default(TopDownAnalyzerFacadeForJVM.kt:96)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:262)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler$analyze$1.invoke(KotlinToJVMBytecodeCompiler.kt:53)
	at org.jetbrains.kotlin.cli.common.messages.AnalyzerWithCompilerReport.analyzeAndReport(AnalyzerWithCompilerReport.kt:113)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.analyze(KotlinToJVMBytecodeCompiler.kt:253)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:100)
	... 24 more
```

--- a/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt
+++ b/compiler-plugin/src/main/kotlin/com/google/devtools/ksp/Incremental.kt
@@ -231,22 +231,27 @@ class IncrementalContext(
         }
     }
 
-    private fun updateLookupCache(dirtyFiles: Collection<File>, removedOutputs: List<File>) {
-        symbolLookupCache.update(
-            symbolLookupTracker, dirtyFiles,
-            options.knownRemoved + removedOutputs.map {
-                it.absoluteFile
-            }
-        )
+    private val removedOutputsKey = File("<This is a virtual key for removed outputs; DO NOT USE>")
+
+    private fun updateFromRemovedOutputs() {
+        val removedOutputs = sourceToOutputsMap.get(removedOutputsKey) ?: return
+
+        symbolLookupCache.removeLookupsFrom(removedOutputs.asSequence())
+        classLookupCache.removeLookupsFrom(removedOutputs.asSequence())
+        removedOutputs.forEach {
+            symbolsMap.remove(it)
+            sealedMap.remove(it)
+        }
+
+        sourceToOutputsMap.remove(removedOutputsKey)
+    }
+
+    private fun updateLookupCache(dirtyFiles: Collection<File>) {
+        symbolLookupCache.update(symbolLookupTracker, dirtyFiles, options.knownRemoved)
         symbolLookupCache.flush(false)
         symbolLookupCache.close()
 
-        classLookupCache.update(
-            classLookupTracker, dirtyFiles,
-            options.knownRemoved + removedOutputs.map {
-                it.absoluteFile
-            }
-        )
+        classLookupCache.update(classLookupTracker, dirtyFiles, options.knownRemoved)
         classLookupCache.flush(false)
         classLookupCache.close()
     }
@@ -272,8 +277,9 @@ class IncrementalContext(
 
         logDirtyFilesByCP(dirtyFilesByCP)
 
+        val removedOutputs = sourceToOutputsMap.get(removedOutputsKey) ?: emptyList()
         // Add previously defined symbols in removed and modified files
-        (modified + removed + dirtyFilesByCP).forEach { file ->
+        (modified + removed + dirtyFilesByCP + removedOutputs).forEach { file ->
             symbolsMap[file]?.let {
                 changedSyms.addAll(it)
             }
@@ -289,7 +295,9 @@ class IncrementalContext(
             invalidator.invalidate(it)
         }
 
-        return invalidator.visitedFiles
+        // visited files are from lookup cache which may contain outputs.
+        val sources = ksFiles.map { it.relativeFile }
+        return invalidator.visitedFiles.filterTo(mutableSetOf()) { it in sources }
     }
 
     // Propagate dirtiness by source-output maps.
@@ -299,7 +307,7 @@ class IncrementalContext(
     ): Set<File> {
         val outputToSources = mutableMapOf<File, MutableSet<File>>()
         sourceToOutputs.keys.forEach { source ->
-            if (source != anyChangesWildcard) {
+            if (source != anyChangesWildcard && source != removedOutputsKey) {
                 sourceToOutputs[source]!!.forEach { output ->
                     outputToSources.getOrPut(output) { mutableSetOf() }.add(source)
                 }
@@ -444,6 +452,8 @@ class IncrementalContext(
                 calcDirtySetByOutputs(sourceToOutputsMap, dirtyFilesByDeps + removed)
             }
 
+            updateFromRemovedOutputs()
+
             logDirtyFilesByOutputs(dirtyFilesByOutputs)
             logDirtyFiles(ksFiles.filter { it.relativeFile in dirtyFilesByOutputs }, ksFiles)
             return ksFiles.filter { it.relativeFile in dirtyFilesByOutputs }
@@ -472,6 +482,7 @@ class IncrementalContext(
         removedOutputs.forEach {
             sourceToOutputsMap.remove(it)
         }
+        sourceToOutputsMap[removedOutputsKey] = removedOutputs
 
         // Update source-to-outputs map from those reprocessed.
         sourceToOutputs.forEach { src, outs ->
@@ -538,7 +549,7 @@ class IncrementalContext(
         val oldOutputs = dirtyFiles.flatMap { sourceToOutputsMap[it] ?: emptyList() }.distinct()
         val removedOutputs = oldOutputs.filterNot { it in outputs }
         updateSourceToOutputs(dirtyFiles, outputs, sourceToOutputs, removedOutputs)
-        updateLookupCache(dirtyFiles, removedOutputs)
+        updateLookupCache(dirtyFiles)
 
         // Update symbolsMap
         fun <K : Comparable<K>, V> update(m: PersistentMap<K, Collection<V>>, u: MultiMap<K, V>) {
@@ -557,10 +568,10 @@ class IncrementalContext(
 
         if (!rebuild) {
             update(sealedMap, updatedSealed)
-            remove(sealedMap, removed + removedOutputs)
+            remove(sealedMap, removed)
 
             update(symbolsMap, updatedSymbols)
-            remove(symbolsMap, removed + removedOutputs)
+            remove(symbolsMap, removed)
         } else {
             symbolsMap.clean()
             update(symbolsMap, updatedSymbols)
@@ -612,7 +623,7 @@ class IncrementalContext(
 
         val cleanOutputs = mutableSetOf<File>()
         sourceToOutputsMap.keys.forEach { source ->
-            if (source !in dirtyFilePaths && source != anyChangesWildcard)
+            if (source !in dirtyFilePaths && source != anyChangesWildcard && source != removedOutputsKey)
                 cleanOutputs.addAll(sourceToOutputsMap[source]!!)
         }
         sourceToOutputsMap.close()
@@ -794,7 +805,7 @@ internal class DepInvalidator(
         if (file in visitedFiles)
             return
         visitedFiles.add(file)
-        symbolsMap[file]!!.forEach {
+        symbolsMap[file]?.forEach {
             invalidate(it)
         }
     }

--- a/integration-tests/src/test/kotlin/com/google/devtools/ksp/test/IncrementalRemovalIT.kt
+++ b/integration-tests/src/test/kotlin/com/google/devtools/ksp/test/IncrementalRemovalIT.kt
@@ -1,0 +1,35 @@
+package com.google.devtools.ksp.test
+
+import org.gradle.testkit.runner.GradleRunner
+import org.junit.Assert
+import org.junit.Rule
+import org.junit.Test
+import java.io.File
+
+class IncrementalRemovalIT {
+    @Rule
+    @JvmField
+    val project: TemporaryTestProject = TemporaryTestProject("incremental-removal")
+
+    @Test
+    fun testRemoveOutputs() {
+        val gradleRunner = GradleRunner.create().withProjectDir(project.root)
+        val k1 = "workload/src/main/kotlin/p1/K1.kt"
+
+        gradleRunner.withArguments("run").build().let { result ->
+            Assert.assertTrue(result.output.contains("result: generated"))
+        }
+
+        File(project.root, k1).writeText(
+            "package p1\n\nclass K1\n\nclass Foo : Bar { override fun s() = \"crafted\" }\n"
+        )
+        gradleRunner.withArguments("run").build().let { result ->
+            Assert.assertTrue(result.output.contains("result: crafted"))
+        }
+
+        project.restore(k1)
+        gradleRunner.withArguments("run").build().let { result ->
+            Assert.assertTrue(result.output.contains("result: generated"))
+        }
+    }
+}

--- a/integration-tests/src/test/resources/incremental-removal/build.gradle.kts
+++ b/integration-tests/src/test/resources/incremental-removal/build.gradle.kts
@@ -1,0 +1,8 @@
+plugins {
+    kotlin("jvm")
+}
+
+repositories {
+    mavenCentral()
+    maven("https://maven.pkg.jetbrains.space/kotlin/p/kotlin/bootstrap/")
+}

--- a/integration-tests/src/test/resources/incremental-removal/gradle.properties
+++ b/integration-tests/src/test/resources/incremental-removal/gradle.properties
@@ -1,0 +1,2 @@
+ksp.incremental=true
+ksp.incremental.log=true

--- a/integration-tests/src/test/resources/incremental-removal/settings.gradle.kts
+++ b/integration-tests/src/test/resources/incremental-removal/settings.gradle.kts
@@ -1,0 +1,19 @@
+pluginManagement {
+    val kspVersion: String by settings
+    val kotlinVersion: String by settings
+    val testRepo: String by settings
+    plugins {
+        id("com.google.devtools.ksp") version kspVersion
+        kotlin("jvm") version kotlinVersion
+    }
+    repositories {
+        maven(testRepo)
+        gradlePluginPortal()
+        maven("https://maven.pkg.jetbrains.space/kotlin/p/kotlin/bootstrap/")
+    }
+}
+
+rootProject.name = "incremental-test"
+
+include(":workload")
+include(":validator")

--- a/integration-tests/src/test/resources/incremental-removal/validator/build.gradle.kts
+++ b/integration-tests/src/test/resources/incremental-removal/validator/build.gradle.kts
@@ -1,0 +1,24 @@
+val kspVersion: String by project
+val testRepo: String by project
+
+plugins {
+    kotlin("jvm")
+}
+
+group = "com.example"
+version = "1.0-SNAPSHOT"
+
+repositories {
+    maven(testRepo)
+    mavenCentral()
+    maven("https://maven.pkg.jetbrains.space/kotlin/p/kotlin/bootstrap/")
+}
+
+dependencies {
+    implementation(kotlin("stdlib"))
+    implementation("com.google.devtools.ksp:symbol-processing-api:$kspVersion")
+}
+
+sourceSets.main {
+    java.srcDirs("src/main/kotlin")
+}

--- a/integration-tests/src/test/resources/incremental-removal/validator/src/main/kotlin/Validator.kt
+++ b/integration-tests/src/test/resources/incremental-removal/validator/src/main/kotlin/Validator.kt
@@ -1,0 +1,48 @@
+import com.google.devtools.ksp.processing.*
+import com.google.devtools.ksp.symbol.*
+import com.google.devtools.ksp.validate
+import java.io.OutputStreamWriter
+
+class Validator : SymbolProcessor {
+    lateinit var codeGenerator: CodeGenerator
+    lateinit var logger: KSPLogger
+
+    fun init(
+        options: Map<String, String>,
+        kotlinVersion: KotlinVersion,
+        codeGenerator: CodeGenerator,
+        logger: KSPLogger,
+    ) {
+        this.codeGenerator = codeGenerator
+        this.logger = logger
+    }
+
+    override fun process(resolver: Resolver): List<KSAnnotated> {
+        resolver.getSymbolsWithAnnotation("p1.MyAnnotation").singleOrNull()?.let { decl ->
+            decl as KSClassDeclaration
+            val file = decl.containingFile!!
+            OutputStreamWriter(
+                codeGenerator.createNewFile(
+                    Dependencies(false, file),
+                    "p1", "Foo"
+                )
+            ).use {
+                it.write("package p1\n\nclass Foo : Bar { override fun s() = \"generated\" }\n")
+            }
+        }
+        resolver.getNewFiles().forEach {
+            it.validate()
+        }
+        return emptyList()
+    }
+}
+
+class TestProcessorProvider : SymbolProcessorProvider {
+    override fun create(
+        env: SymbolProcessorEnvironment,
+    ): SymbolProcessor {
+        return Validator().apply {
+            init(env.options, env.kotlinVersion, env.codeGenerator, env.logger)
+        }
+    }
+}

--- a/integration-tests/src/test/resources/incremental-removal/validator/src/main/resources/META-INF/services/com.google.devtools.ksp.processing.SymbolProcessorProvider
+++ b/integration-tests/src/test/resources/incremental-removal/validator/src/main/resources/META-INF/services/com.google.devtools.ksp.processing.SymbolProcessorProvider
@@ -1,0 +1,1 @@
+TestProcessorProvider

--- a/integration-tests/src/test/resources/incremental-removal/workload/build.gradle.kts
+++ b/integration-tests/src/test/resources/incremental-removal/workload/build.gradle.kts
@@ -1,0 +1,27 @@
+val testRepo: String by project
+
+plugins {
+    id("com.google.devtools.ksp")
+    kotlin("jvm")
+    application
+}
+
+version = "1.0-SNAPSHOT"
+
+repositories {
+    maven(testRepo)
+    mavenCentral()
+    maven("https://maven.pkg.jetbrains.space/kotlin/p/kotlin/bootstrap/")
+}
+
+dependencies {
+    implementation(kotlin("stdlib"))
+    implementation(project(":validator"))
+    testImplementation("junit:junit:4.12")
+    ksp(project(":validator"))
+    kspTest(project(":validator"))
+}
+
+application {
+    mainClassName = "p1.MainKt"
+}

--- a/integration-tests/src/test/resources/incremental-removal/workload/src/main/kotlin/p1/K1.kt
+++ b/integration-tests/src/test/resources/incremental-removal/workload/src/main/kotlin/p1/K1.kt
@@ -1,0 +1,4 @@
+package p1
+
+@MyAnnotation
+class K1

--- a/integration-tests/src/test/resources/incremental-removal/workload/src/main/kotlin/p1/Main.kt
+++ b/integration-tests/src/test/resources/incremental-removal/workload/src/main/kotlin/p1/Main.kt
@@ -1,0 +1,12 @@
+package p1
+
+interface Bar {
+    fun s(): String
+}
+val bar: Foo = Foo()
+
+annotation class MyAnnotation
+
+fun main() {
+    println("result: ${bar.s()}")
+}

